lib/pages/timetable/model/academic_calendar.dart

```diff
@@ -77,4 +78,45 @@ class AcademicCalendar {
 
     return weeksByYear.values.expand((e) => e).toSet();
   }
+
+  int semesterForDate(LocalDate date) {
+    for (final semester in semesters) {
+      if (date._isBeforeOrDuring(semester)) {
+        return 1 + int.tryParse(semester.id[semester.id.length - 1]);
+      }
+    }
+    return -1;
+  }
+
+  String getWeekNumber(LocalDate date) {
+    final week = ((date.dayOfYear - date.dayOfWeek.value + 10) / 7).floor();
+    if (LeadHeader.academicWeekNumber == false) return week.toString();
+    final int finalWeekOfFirstSem = ((semesters[0].endDate.dayOfYear -
+                semesters[0].endDate.dayOfWeek.value +
+                10) /
+            7)
+        .floor();
+    if (!nonHolidayWeeks.contains(week)) {
+      return 'H';
+    } else {
+      if (semesterForDate(date) == 1) {
+        return (nonHolidayWeeks.toList().indexOf(week) + 1).toString();
+      } else {
+        return ((nonHolidayWeeks.toList().indexOf(week) + 1) -
```

Comment:
I don't like this logic, you're kind of assuming that the holiday between the semesters is exactly one week. This is very likely true for ACS, but if we ever want to extend this, it will lead to a bug that will be tricky to figure out. Similarly, you're also assuming that we have only two semesters :) This is meant to be a generic API.

Instead of calculating the last week of the first semester, why not just get the first week of the each semester, find the index of that in `nonHolidayWeeks` and simply subtract that from `.indexOf(week)`? This would be a generic solution that works for any N semesters with arbitrary holidays.